cloudify/workflows/tasks_graph.py

```diff
@@ -466,18 +454,33 @@ def add_dependency(self, src_task, dst_task):
         self.graph.add_dependency(src_task, dst_task)
 
     def apply_async(self):
+        super(SubgraphTask, self).apply_async()
         if not self.tasks:
             self.set_state(tasks.TASK_SUCCEEDED)
         else:
+            for task_id, task in self.tasks.items():
+                self.graph.remove_dependency(task, self)
```

Comment:
I may be misreading this, but it looks a bit like we claim the dependency is resolved the moment we start a task rather than on finishing it?

Reply:
yeah. As I noted in one of the commits, it used to be that the "task runs after its graph" was encoded somewhere in the `_executable_task` stack of `if`s. Now it's directed by the dependency that tasks have on their subgraphs. So when a subgraph starts, we remove that dependency, so that the tasks can run.

This is as opposed to finishing the subgraph right there, because we do want the subgraph to only finish when its tasks finish.

An alternative would be additional `if`s inside of the graph itself. This approach turned out to be cleaner.

Also added a comment to the code saying as much

Reply:
But doesn't this mean that something that depends on task x will (be able to) start running as soon as task x starts, without waiting for it to finish?

That could presumably result in something failing or needing to retry due to the actions task x should take not yet being complete?

Reply:
you're reading the implementation of a SubgraphTask here, so yes, but only if task x is a subgraph.

Of course, tasks inside of the subgraph can and will have also other dependencies that they will have to wait for: eg. in a subgraph S containing a sequence of tasks [A, B, C]:
  - task A has dependencies on S,
  - task B has dependencies on S, A
  - task C has dependencies on S, B
so starting the subgraph resolves that S dependency, allowing A to run, but B will still wait for A (but not for S anymore)

Reply:
Ah, ok, that makes sense. Marking as resolved then.